### PR TITLE
Block temporary/burner email addresses

### PR DESCRIPTION
**Pull request by jeffin143**
Handling : https://github.com/opencollective/opencollective/issues/2576

Screenshot from 2019-10-31 19-57-18(1)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,6 +29,7 @@
     "bcrypt": "3.0.6",
     "bluebird": "3.7.1",
     "body-parser": "1.19.0",
+    "burner-email-providers": "^1.0.6",
```
Comment (Betree) [inline package.json]: We prefer to use fixed versions to better detect breaking changes

```suggestion
    "burner-email-providers": "1.0.6",
```

File: server/models/User.js [hunk 2]
```diff
@@ -48,6 +49,13 @@ export default (Sequelize, DataTypes) => {
           isEmail: {
             msg: 'Email must be valid',
           },
+          isfake: function(val) {
```
Comment (Betree) [inline server/models/User.js]: May be more precise

```suggestion
          isBurnerEmail: function(val) {
```

**Comment by znarf** [comment]
@effin143 Can you rebase from master? Then look at the failing tests?

**Comment by znarf** [comment]
@Betree I think this needs to be documented, we should have a clear idea of what is allowed or not. Would be also interesting to run a test on already registered emails to check that it's making sense.

**Comment by Betree** [comment]
Thanks for pointing that out @znarf, from the extracted data I now think that we should have a small whitelist of domains we want to allow. 

For example https://hush.com  and https://mailhero.io seem like legitimate services that are used to avoid SPAM. They are **private** emails providers, thus they don't fall in the initial description of shared public inboxes like http://www.yopmail.com/ (that is often used by robots because they make automatizing emails reception really easy).

Below is the data. My advice would be to **whitelist all services from `Blocked providers for Users with at least 1 transaction` that are not shared public inboxes**.

---

Number of existing users: **1157**
Number of existing users with at least one transaction: **464**

<details>
<summary>Blocked providers for Users with at least 1 transaction</summary>

|          Domain                 |   Count  | 
|---------------------------|-----| 
| https://hush.com          | 100 | 
| https://opayq.com         | 94  | 
| https://mailhero.io       | 70  | 
| https://mailinator.com    | 46  | 
| https://sudomail.com      | 34  | 
| https://maildrop.cc       | 32  | 
| https://neverbox.com      | 24  | 
| https://thrott.com        | 18  | 
| https://gmial.com         | 12  | 
| https://sharklasers.com   | 4   | 
| https://spamgourmet.com   | 4   | 
| https://guerrillamail.com | 2   | 
| https://urhen.com         | 2   | 
| https://max-mail.com      | 2   | 
| https://daymail.life      | 2   | 
| https://cndps.com         | 2   | 
| https://nicemail.pro      | 2   | 
| https://its0k.com         | 2   | 
| https://spamex.com        | 2   | 
| https://zebra.email       | 2   | 
| https://byom.de           | 2   | 
| https://yeah.net          | 2   | 
| https://nigge.rs          | 2   | 
| https://xoxy.net          | 2   | 

</details>

<details>
<summary>All blocked providers having entries in Users</summary>

```
mailprotech.com
simpleemail.info
i.ua
sharklasers.com
spamex.com
opayq.com
theroyalweb.club
wegwerfemail.de
yeah.net
mvrht.net
sudomail.com
crypemail.info
bobmail.info
dispostable.com
dreamcatcher.email
urhen.com
skymailapp.com
gmal.com
nigge.rs
planet-travel.club
guerrillamail.com
nwytg.com
next-mail.info
oranek.com
nezdiro.org
loketa.com
hidebusiness.xyz
gamil.com
max-mail.com
yk20.com
4senditnow.com
patonce.com
mailbox87.de
hotmai.com
dayrep.com
zebra.email
mygeoweb.info
minex-coin.com
geo-crypto.com
zoqqa.com
maileme101.com
69postix.info
mailinator.com
binka.me
mailbox52.ga
thrott.com
hitbts.com
getnada.com
test.com
cowcell.com
gmial.com
alexbox.online
mail4-us.org
trash-mail.com
furusato.tokyo
polyswarms.com
mvrht.com
dnsdeer.com
geroev.net
hush.com
neko2.net
example.com
grr.la
shop4mail.net
neverbox.com
etoic.com
bossmail.de
mohmal.com
daymail.life
vomoto.com
its0k.com
nwytg.net
himail.online
xoxy.net
bcaoo.com
f5.si
smartbusiness.me
nicemail.pro
jmail.ovh
mailnesia.com
reddcoin2.com
crowd-mail.com
eoffice.top
maildrop.cc
byom.de
umail2.com
vpsorg.top
honor-8.com
armyspy.com
emailna.co
mailtothis.com
mt2015.com
safetymail.info
spamgourmet.com
fastair.info
desoz.com
theskymail.com
bccto.me
rockmailapp.com
mail-cart.com
lavabit.com
pay-mon.com
phpbb.uu.gl
ymail4.com
hideweb.xyz
4mail.top
cndps.com
mailhero.io
macr2.com
yopmail.com
ugimail.net
alltempmail.com
```

</details>

**Comment by znarf** [comment]
> Below is the data. My advice would be to whitelist all services from Blocked providers for Users with at least 1 transaction that are not shared public inboxes.

@Betree I don't see that implemented in the PR. Shouldn't that be tackled?

**Comment by jeffin143** [comment]
> @znarf Yes, specs from [#2832 (comment)](https://github.com/opencollective/opencollective-api/pull/2832#issuecomment-553325670) are a blocker before we can merge this.

How should I implement it 

 - make a text file with all the domains
 -  if domain is present in that list , allow them

**Comment by Betree** [comment]
> > @znarf Yes, specs from [#2832 (comment)](https://github.com/opencollective/opencollective-api/pull/2832#issuecomment-553325670) are a blocker before we can merge this.
> 
> How should I implement it
> 
>     * make a text file with all the domains
>     * if domain is present in that list , allow them

Yes, except it should be a JS file - not a text file. The most appropriate one to add this for me seems to be `server/lib/email.js`. We can have a `isWhitelistedDomain` function like this in it:

```es6
const isWhitelistedDomain = domain => {
  return [].includes(domain) // TODO add domains here
}
```

File: server/lib/email.js [hunk 3]
```diff
@@ -226,6 +227,11 @@ const getNotificationLabel = (template, recipients) => {
   return notificationTypeLabels[template];
 };
 
+const isWhitelistedDomain = email => {
+  const domain = email.toString().split('@');
```
Comment (Betree) [inline server/lib/email.js]: I'm not sure we need that, do we?

```suggestion
  const domain = email.split('@');
```

Comment (jeffin143) [reply server/lib/email.js]: Made changes :) 

It was not needed 😊

**Comment by znarf** [comment]
@Betree can you do a final review and approve?

**Pull request merged**